File: flink-ml-python/pyflink/examples/ml/recommendation/fpgrowth_example.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################

# Simple program that creates a fpgrowth instance and gives recommendations for items.

from pyflink.common import Types
from pyflink.datastream import StreamExecutionEnvironment
from pyflink.table import StreamTableEnvironment

from pyflink.ml.recommendation.fpgrowth import FPGrowth

# Creates a new StreamExecutionEnvironment.
env = StreamExecutionEnvironment.get_execution_environment()

# Creates a StreamTableEnvironment.
t_env = StreamTableEnvironment.create(env)

# Generates input data.
input_table = t_env.from_data_stream(
    env.from_collection([
        ("A,B,C,D",),
        ("B,C,E",),
        ("A,B,C,E",),
        ("B,D,E",),
        ("A,B,C,D",)
    ],
        type_info=Types.ROW_NAMED(
            ['items'],
            [Types.STRING()])
    ))

# Creates a fpgrowth object and initialize its parameters.
fpg = FPGrowth().set_min_support(0.6)

# Transforms the data to fpgrowth algorithm result.
output_table = fpg.transform(input_table)

# Extracts and display the results.
pattern_result_names = output_table[0].get_schema().get_field_names()
rule_result_names = output_table[1].get_schema().get_field_names()

patterns = t_env.to_data_stream(output_table[0]).execute_and_collect()
rules = t_env.to_data_stream(output_table[1]).execute_and_collect()

print("|\t"+"\t|\t".join(pattern_result_names)+"\t|")
for result in patterns:
    print(f'|\t{result[0]}\t|\t{result[1]}\t|\t{result[2]}\t|')
print("|\t"+" | ".join(rule_result_names)+"\t|")
for result in rules:
    print(f'|\t{result[0]}\t|\t{result[1]}\t|\t{result[2]}\t|\t{result[3]}'
          + f'\t|\t{result[4]}\t|\t{result[5]}\t|')

File: flink-ml-examples/src/main/java/org/apache/flink/ml/examples/recommendation/FPGrowthExample.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.ml.examples.recommendation;

import org.apache.flink.ml.recommendation.fpgrowth.FPGrowth;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
import org.apache.flink.types.Row;
import org.apache.flink.util.CloseableIterator;

/**
 * Simple program that creates a FPGrowth instance and uses it to generate frequent patterns and
 * association rules.
 */
public class FPGrowthExample {
    public static void main(String[] args) {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        StreamTableEnvironment tEnv = StreamTableEnvironment.create(env);

        // Generates input data.
        DataStream<Row> inputStream =
                env.fromElements(
                        Row.of(""),
                        Row.of("A,B,C,D"),
                        Row.of("B,C,E"),
                        Row.of("A,B,C,E"),
                        Row.of("B,D,E"),
                        Row.of("A,B,C,D,A"));

        Table inputTable = tEnv.fromDataStream(inputStream).as("items");

        // Creates a FPGrowth object and initializes its parameters.
        FPGrowth fpg = new FPGrowth().setMinSupportCount(3);

        // Transforms the data.
        Table[] outputTable = fpg.transform(inputTable);

        // Extracts and displays the frequent patterns.
        for (CloseableIterator<Row> it = outputTable[0].execute().collect(); it.hasNext(); ) {
            Row row = it.next();

            String pattern = row.getFieldAs(0);
            Long support = row.getFieldAs(1);
            Long itemCount = row.getFieldAs(2);

            System.out.printf(
                    "pattern: %s, support count: %d, item_count:%d\n", pattern, support, itemCount);
        }

        // Extracts and displays the association rules.
        for (CloseableIterator<Row> it = outputTable[1].execute().collect(); it.hasNext(); ) {
            Row row = it.next();

            String rule = row.getFieldAs(0);
            Double lift = row.getFieldAs(2);
            Double support = row.getFieldAs(3);
            Double confidence_percent = row.getFieldAs(4);

            System.out.printf(
                    "rule: %s, list: %f, support:%f, confidence:%f\n",
                    rule, lift, support, confidence_percent);
        }
    }
}
